torch_brain/models/poyo.py

```diff
@@ -0,0 +1,354 @@
+        (
+            se_token_type_index,
+            se_unit_index,
+            se_timestamps,
+        ) = create_start_end_unit_tokens(unit_ids, start, end)
+
+        # append start and end tokens to the spike sequence
+        spike_token_type_index = np.concatenate(
+            [se_token_type_index, np.zeros_like(spike_unit_index)]
+        )
+        spike_unit_index = np.concatenate([se_unit_index, spike_unit_index])
+        spike_timestamps = np.concatenate([se_timestamps, spike_timestamps])
+
+        # unit_index is relative to the recording, so we want it to map it to
+        # the global unit index
+        local_to_global_map = np.array(self.unit_tokenizer(unit_ids))
+        spike_unit_index = local_to_global_map[spike_unit_index]
+
+        ### prepare latents
+        latent_index, latent_timestamps = create_linspace_latent_tokens(
+            start,
+            end,
+            step=self.latent_step,
+            num_latents_per_step=self.num_latents_per_step,
+        )
+
+        ### prepare output queries and targets
+        output_timestamps = data.get_nested_attribute(self.modality_spec.timestamp_key)
+        output_values = data.get_nested_attribute(self.modality_spec.value_key)
+        if output_values.dtype == np.float64:
+            output_values = output_values.astype(np.float32)
+
+        session_index = self.session_tokenizer(data.session)
+        session_index = np.repeat(session_index, len(output_timestamps))
+
+        # Weights for the output predictions (used in the loss function)
+        output_subtask_index = data.get_nested_attribute(self.modality_spec.context_key)
+        if self.subtask_weights is None:
+            output_weights = np.ones(len(output_values), dtype=np.float32)
+        else:
+            output_weights = self.subtask_weights[output_subtask_index]
+
+        # Mask for the output predictions
+        output_mask = np.ones(len(output_values), dtype=bool)
+        if self.eval:
+            # During eval, only evaluate on the subtask specified in the config
+            target_subtask_index = data.config["eval_subtask_index"]
+            output_mask = output_mask & (output_subtask_index == target_subtask_index)
+
+        batch = {
+            # input sequence
+            "input_unit_index": pad8(spike_unit_index),
+            "input_timestamps": pad8(spike_timestamps),
+            "input_token_type": pad8(spike_token_type_index),
+            "input_mask": track_mask8(spike_unit_index),
+            # latent sequence
+            "latent_index": latent_index,
+            "latent_timestamps": latent_timestamps,
+            # output sequence
+            "output_session_index": pad8(session_index),
+            "output_timestamps": pad8(output_timestamps),
+            "output_mask": pad8(output_mask),
+            # ground truth targets
+            "target_values": pad8(output_values),
+            "target_weights": pad8(output_weights),
```

Comment:
here padding is used instead of chaining. how are you dealing with the loss when it comes to padding values?